File: golang/datastructure/queue/priorityqueue/priorityqueue.go
package priorityqueue

type PriorityQueue struct {
	q              []interface{}
	size, capacity int
	Less           func(a, b interface{}) bool
}

func CreatePriorityQueue(n int, less func(a, b interface{}) bool) PriorityQueue {
	return PriorityQueue{make([]interface{}, n), 0, n, less}
}

func (pq *PriorityQueue) Push(v interface{}) bool {
	if pq.IsFull() {
		return false
	}
	pq.q[pq.size] = v
	pq.size++
	pq.up(pq.size - 1)
	return true
}

func (pq *PriorityQueue) Pop() interface{} {
	if pq.IsEmpty() {
		return nil
	}
	top := pq.q[0]
	pq.q[0] = pq.q[pq.size-1]
	pq.size--
	pq.down(0)
	return top
}

func (pq *PriorityQueue) Top() interface{} {
	return pq.q[0]
}

func (pq *PriorityQueue) IsFull() bool {
	return pq.size == pq.capacity
}

func (pq *PriorityQueue) IsEmpty() bool {
	return pq.size == 0
}

func (pq *PriorityQueue) up(pos int) {
	for pos > 0 {
		i := (pos - 1) / 2
		if pq.less(i, pos) {
			return
		}
		pq.q[i], pq.q[pos] = pq.q[pos], pq.q[i]
		pos = i
	}
}

func (pq *PriorityQueue) down(pos int) {
	for pos < pq.size {
		small, l := pos, 2*pos+1
		if l < pq.size && pq.less(l, small) {
			small = l
		}
		r := l + 1
		if r < pq.size && pq.less(r, small) {
			small = r
		}
		if small == pos {
			return
		}
		pq.q[pos], pq.q[small] = pq.q[small], pq.q[pos]
		pos = small
	}
}

func (pq *PriorityQueue) less(i, j int) bool {
	a, b := pq.q[i], pq.q[j]
	return pq.Less(a, b)
}

File: c/datastructure/queue/priorityqueue/priorityqueue.c
//
// Created by john on 2022/9/30.
//
#include "priorityqueue.h"

bool isEmpty(PriorityQueue *pq) {
    return pq->size == 0;
}

bool isFull(PriorityQueue *pq) {
    return pq->size == pq->capacity;
}

void *top(PriorityQueue *pq) {
    if (isEmpty(pq)) {
        return NULL;
    }
    return pq->q[0];
}

void swap(PriorityQueue *pq, int i, int j) {
    void *t = pq->q[i];
    pq->q[i] = pq->q[j];
    pq->q[j] = t;
}

bool less(PriorityQueue *pq, int i, int j) {
    return pq->less(pq->q[i], pq->q[j]);
}

void up(PriorityQueue *pq, int p) {
    while (p > 0) {
        int r = (p - 1) / 2;
        if (less(pq, r, p)) {
            return;
        }
        swap(pq, r, p);
        p = r;
    }
}

void down(PriorityQueue *pq, int p) {
    while (p < pq->size) {
        int small = p, l = 2 * p + 1;
        if (l < pq->size && less(pq, l, small)) {
            small = l;
        }
        int r = l + 1;
        if (r < pq->size && less(pq, r, small)) {
            small = r;
        }
        if (small == p) {
            return;
        }
        swap(pq, small, p);
        p = small;
    }
}

bool push(PriorityQueue *pq, void *t) {
    if (isFull(pq)) {
        return false;
    }
    pq->q[pq->size] = t;
    pq->size++;
    up(pq, pq->size - 1);
    return true;
}

void *pop(PriorityQueue *pq) {
    if (isEmpty(pq)) {
        return NULL;
    }
    void *top = pq->q[0];
    pq->size--;
    pq->q[0] = pq->q[pq->size];
    down(pq, 0);
    return top;
}


PriorityQueue createPriorityQueue(int n, bool (*less)(void *, void *)) {
    PriorityQueue pq;
    pq.size = 0;
    pq.capacity = n;
    pq.q = (void **) malloc(n * sizeof(void *));
    pq.less = less;
    return pq;
}

void releasePriorityQueue(PriorityQueue *pq) {
    if (pq) {
        free(pq->q);
        pq->q = NULL;
    }
}